File: packages/core-python/payafrica/providers/wave.py
from __future__ import annotations

import hashlib
import hmac
import json
from typing import Any

import httpx

from ..contracts import PaymentProvider, WebhookHeaders
from ..enums import PaymentError, PaymentStatus
from ..errors import ProviderError
from ..models import PaymentEvent, PaymentRequest, PaymentSession, RefundResult


class WaveProvider(PaymentProvider):
    _BASE_URL = "https://api.wave.com/v1"

    def __init__(self, client: httpx.AsyncClient, api_key: str, webhook_secret: str) -> None:
        self._client, self._api_key, self._webhook_secret = client, api_key, webhook_secret

    async def initiate_payment(self, params: PaymentRequest) -> PaymentSession:
        payload = await self._request("POST", "/checkout/sessions", json={"amount": params.amount, "currency": "XOF", "error_url": params.failure_url, "success_url": params.success_url, "client_reference": params.reference})
        session_id = payload.get("id")
        if not isinstance(session_id, str): raise ProviderError(PaymentError.UNKNOWN, "Wave checkout response is missing id")
        return PaymentSession(id=session_id, reference=params.reference, amount=params.amount, currency=params.currency, status=PaymentStatus.PENDING, payment_url=payload.get("wave_launch_url") if isinstance(payload.get("wave_launch_url"), str) else None)

    async def check_status(self, session_id: str) -> PaymentStatus:
        return self._status((await self._request("GET", f"/checkout/sessions/{session_id}" )).get("payment_status"))

    async def handle_webhook(self, raw_body: str | bytes, headers: WebhookHeaders) -> PaymentEvent:
        raw = raw_body.encode() if isinstance(raw_body, str) else raw_body
        signature = self._header(headers, "x-wave-signature")
        expected = hmac.new(self._webhook_secret.encode(), raw, hashlib.sha256).hexdigest()
        candidates = [part.strip().removeprefix("v1=") for part in signature.split(",")] if signature else []
        if not any(hmac.compare_digest(expected, candidate) for candidate in candidates): raise ProviderError(PaymentError.UNKNOWN, "Invalid Wave webhook signature")
        payload = self._decode(raw); data = payload.get("data")
        if not isinstance(data, dict) or not isinstance(data.get("id"), str): raise ProviderError(PaymentError.UNKNOWN, "Incomplete Wave webhook payload")
        event_type = payload.get("type"); status = PaymentStatus.SUCCESS if event_type == "checkout.session.completed" else PaymentStatus.FAILED if event_type == "checkout.session.payment_failed" else self._status(data.get("payment_status"))
        return PaymentEvent(id=str(payload.get("id", data["id"])), session_id=data["id"], status=status, occurred_at=str(data.get("when_completed", data.get("when_created", "1970-01-01T00:00:00Z"))), reference=data.get("client_reference") if isinstance(data.get("client_reference"), str) else None)

    async def refund(self, session_id: str, amount: int | None = None) -> RefundResult:
        payload = await self._request("POST", f"/checkout/sessions/{session_id}/refund", json={} if amount is None else {"amount": amount})
        refund_id, refund_amount = payload.get("id"), payload.get("amount")
        if not isinstance(refund_id, str) or not isinstance(refund_amount, int): raise ProviderError(PaymentError.UNKNOWN, "Incomplete Wave refund response")
        return RefundResult(session_id=session_id, refund_id=refund_id, amount=refund_amount, status=self._status(payload.get("status", "succeeded")))

    async def _request(self, method: str, path: str, **kwargs: Any) -> dict[str, Any]:
        try: response = await self._client.request(method, self._BASE_URL + path, headers={"Authorization": f"Bearer {self._api_key}"}, **kwargs)
        except httpx.HTTPError as exc: raise ProviderError(PaymentError.PROVIDER_TIMEOUT, str(exc)) from exc
        payload = response.json() if response.content else {}; payload = payload if isinstance(payload, dict) else {}
        if response.is_error:
            code = str(payload.get("error_code", "")); error = PaymentError.INSUFFICIENT_FUNDS if code == "insufficient-funds" else PaymentError.INVALID_PHONE if code in {"payer-mobile-mismatch", "invalid-phone"} else PaymentError.PROVIDER_TIMEOUT if response.status_code >= 500 or response.status_code == 408 else PaymentError.UNKNOWN
            raise ProviderError(error, str(payload.get("error_message", "Wave request failed")))
        return payload

    @staticmethod
    def _status(status: object) -> PaymentStatus:
        return {"succeeded": PaymentStatus.SUCCESS, "success": PaymentStatus.SUCCESS, "processing": PaymentStatus.PENDING, "pending": PaymentStatus.PENDING, "cancelled": PaymentStatus.FAILED, "failed": PaymentStatus.FAILED}.get(str(status).lower(), PaymentStatus.FAILED)

    @staticmethod
    def _decode(raw: bytes) -> dict[str, Any]:
        payload = json.loads(raw)
        if not isinstance(payload, dict): raise ProviderError(PaymentError.UNKNOWN, "Invalid Wave webhook payload")
        return payload

    @staticmethod
    def _header(headers: WebhookHeaders, name: str) -> str | None:
        return next((value for key, value in headers.items() if key.lower() == name and isinstance(value, str)), None)

File: packages/core-node/src/providers/wave.ts
import { createHmac, timingSafeEqual } from "node:crypto";

import {
  PaymentError,
  PaymentStatus,
  type PaymentEvent,
  type PaymentProvider,
  type PaymentRequest,
  type PaymentSession,
  type RefundResult,
} from "../types.js";

const WAVE_API_BASE_URL = "https://api.wave.com/v1";

export interface WaveProviderConfig {
  apiKey: string;
  webhookSecret: string;
}

interface WaveCheckoutSession {
  id?: string;
  amount?: string | number;
  currency?: string;
  client_reference?: string | null;
  wave_launch_url?: string;
  payment_status?: string;
}

interface WaveRefundResponse {
  id?: string;
  amount?: string | number;
  status?: string;
}

interface WaveErrorResponse {
  error_code?: string;
  error_message?: string;
  message?: string;
}

interface WaveWebhookPayload {
  id?: string;
  type?: string;
  data?: WaveCheckoutSession & { when_completed?: string; when_created?: string };
}

/** Error intentionally reduced to the common PaymentError code. */
export class WaveProviderError extends Error {
  public constructor(public readonly code: PaymentError, message: string) {
    super(message);
    this.name = "WaveProviderError";
  }
}

export class WaveProvider implements PaymentProvider {
  public constructor(private readonly config: WaveProviderConfig) {}

  public async initiatePayment(params: PaymentRequest): Promise<PaymentSession> {
    const response = await this.request("/checkout/sessions", {
      method: "POST",
      body: JSON.stringify({
        amount: params.amount,
        currency: "XOF",
        ...(params.failureUrl === undefined ? {} : { error_url: params.failureUrl }),
        ...(params.successUrl === undefined ? {} : { success_url: params.successUrl }),
        client_reference: params.reference,
      }),
    });
    const session = await this.readJson<WaveCheckoutSession>(response);

    if (typeof session.id !== "string" || session.id.length === 0) {
      throw new WaveProviderError(PaymentError.Unknown, "Wave checkout response is missing id");
    }

    return {
      id: session.id,
      reference: params.reference,
      amount: params.amount,
      currency: params.currency,
      status: PaymentStatus.Pending,
      ...(session.wave_launch_url === undefined ? {} : { paymentUrl: session.wave_launch_url }),
    };
  }

  public async checkStatus(sessionId: string): Promise<PaymentStatus> {
    const response = await this.request(`/checkout/sessions/${encodeURIComponent(sessionId)}`);
    const session = await this.readJson<WaveCheckoutSession>(response);

    if (typeof session.payment_status !== "string") {
      throw new WaveProviderError(PaymentError.Unknown, "Wave checkout response is missing payment_status");
    }
    return this.toPaymentStatus(session.payment_status);
  }

  public async handleWebhook(
    rawBody: string | Buffer,
    headers: Record<string, string | string[] | undefined>
  ): Promise<PaymentEvent> {
    const signature = this.findHeader(headers, "x-wave-signature");
    const rawBytes = typeof rawBody === "string" ? Buffer.from(rawBody) : rawBody;

    if (signature === undefined || !this.isValidSignature(rawBytes, signature)) {
      throw new WaveProviderError(PaymentError.Unknown, "Invalid Wave webhook signature");
    }

    const payload = this.parseWebhookPayload(rawBytes.toString("utf8"));
    const data = payload.data;
    const sessionId = data?.id;
    if (sessionId === undefined || sessionId.length === 0) {
      throw new WaveProviderError(PaymentError.Unknown, "Wave webhook is missing checkout session id");
    }

    return {
      id: payload.id ?? sessionId,
      sessionId,
      status: this.toWebhookStatus(payload.type, data?.payment_status),
      ...(data?.client_reference === null || data?.client_reference === undefined
        ? {}
        : { reference: data.client_reference }),
      occurredAt: data?.when_completed ?? data?.when_created ?? new Date().toISOString(),
    };
  }

  public async refund(sessionId: string, amount?: number): Promise<RefundResult> {
    const response = await this.request(`/checkout/sessions/${encodeURIComponent(sessionId)}/refund`, {
      method: "POST",
      ...(amount === undefined ? {} : { body: JSON.stringify({ amount }) }),
    });
    const refund = await this.readJson<WaveRefundResponse>(response);

    if (typeof refund.id !== "string" || typeof refund.amount !== "number") {
      throw new WaveProviderError(PaymentError.Unknown, "Wave refund response is incomplete");
    }

    return {
      sessionId,
      refundId: refund.id,
      amount: refund.amount,
      status: refund.status === undefined ? PaymentStatus.Success : this.toPaymentStatus(refund.status),
    };
  }

  private async request(path: string, init: RequestInit = {}): Promise<Response> {
    let response: Response;
    try {
      response = await fetch(`${WAVE_API_BASE_URL}${path}`, {
        ...init,
        headers: {
          Authorization: `Bearer ${this.config.apiKey}`,
          "Content-Type": "application/json",
          ...init.headers,
        },
      });
    } catch (error: unknown) {
      throw this.toProviderError(error);
    }

    if (!response.ok) {
      throw await this.toResponseError(response);
    }
    return response;
  }

  private async toResponseError(response: Response): Promise<WaveProviderError> {
    const body = await this.readJsonOrEmpty<WaveErrorResponse>(response);
    const message = body.error_message ?? body.message ?? `Wave request failed with HTTP ${response.status}`;
    return new WaveProviderError(this.mapError(response.status, body.error_code), message);
  }

  private toProviderError(error: unknown): WaveProviderError {
    if (error instanceof WaveProviderError) {
      return error;
    }
    return new WaveProviderError(
      PaymentError.ProviderTimeout,
      error instanceof Error ? error.message : "Wave request failed"
    );
  }

  private mapError(httpStatus: number, waveCode: string | undefined): PaymentError {
    if (waveCode === "insufficient-funds") {
      return PaymentError.InsufficientFunds;
    }
    if (waveCode === "payer-mobile-mismatch" || waveCode === "invalid-phone") {
      return PaymentError.InvalidPhone;
    }
    if (waveCode === "payment-cancelled" || waveCode === "user-cancelled") {
      return PaymentError.UserCancelled;
    }
    if (httpStatus === 408 || httpStatus === 429 || httpStatus >= 500) {
      return PaymentError.ProviderTimeout;
    }
    return PaymentError.Unknown;
  }

  private toPaymentStatus(status: string): PaymentStatus {
    switch (status.toLowerCase()) {
      case "succeeded":
      case "success":
        return PaymentStatus.Success;
      case "processing":
      case "pending":
        return PaymentStatus.Pending;
      case "cancelled":
      case "failed":
        return PaymentStatus.Failed;
      default:
        throw new WaveProviderError(PaymentError.Unknown, "Unknown Wave payment status");
    }
  }

  private toWebhookStatus(type: string | undefined, paymentStatus: string | undefined): PaymentStatus {
    if (type === "checkout.session.completed") {
      return PaymentStatus.Success;
    }
    if (type === "checkout.session.payment_failed") {
      return PaymentStatus.Failed;
    }
    if (paymentStatus === undefined) {
      throw new WaveProviderError(PaymentError.Unknown, "Wave webhook has an unknown event type");
    }
    return this.toPaymentStatus(paymentStatus);
  }

  private findHeader(
    headers: Record<string, string | string[] | undefined>,
    expectedName: string
  ): string | undefined {
    for (const [name, value] of Object.entries(headers)) {
      if (name.toLowerCase() !== expectedName) {
        continue;
      }
      if (typeof value === "string") {
        return value;
      }
      if (Array.isArray(value) && value.length === 1) {
        return value[0];
      }
    }
    return undefined;
  }

  private isValidSignature(rawBody: Buffer, headerValue: string): boolean {
    const expectedSignature = createHmac("sha256", this.config.webhookSecret).update(rawBody).digest("hex");
    const candidates = headerValue.split(",").map((part) => part.trim().replace(/^v1=/, ""));

    return candidates.some((candidate) => {
      const expected = Buffer.from(expectedSignature);
      const received = Buffer.from(candidate);
      return expected.length === received.length && timingSafeEqual(expected, received);
    });
  }

  private parseWebhookPayload(serializedBody: string): WaveWebhookPayload {
    try {
      const payload: unknown = JSON.parse(serializedBody);
      if (typeof payload !== "object" || payload === null || Array.isArray(payload)) {
        throw new Error("Webhook payload must be an object");
      }
      return payload as WaveWebhookPayload;
    } catch {
      throw new WaveProviderError(PaymentError.Unknown, "Invalid Wave webhook payload");
    }
  }

  private async readJson<T>(response: Response): Promise<T> {
    const body: unknown = await response.json();
    return body as T;
  }

  private async readJsonOrEmpty<T>(response: Response): Promise<T> {
    try {
      return await this.readJson<T>(response);
    } catch {
      return {} as T;
    }
  }
}
